src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -5119,23 +5145,47 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
             }
             else
             {
-                TypeWithState rightType;
+                TypeWithState rightState;
                 if (!node.IsRef)
                 {
-                    rightType = VisitOptionalImplicitConversion(right, leftLValueType, UseLegacyWarnings(left, leftLValueType), trackMembers: true, AssignmentKind.Assignment);
+                    rightState = VisitOptionalImplicitConversion(right, leftLValueType, UseLegacyWarnings(left, leftLValueType), trackMembers: true, AssignmentKind.Assignment);
                 }
                 else
                 {
-                    rightType = VisitRefExpression(right, leftLValueType);
+                    rightState = VisitRefExpression(right, leftLValueType);
                 }
 
-                TrackNullableStateForAssignment(right, leftLValueType, MakeSlot(left), rightType, MakeSlot(right));
-                SetResult(node, TypeWithState.Create(leftLValueType.Type, rightType.State), leftLValueType);
+                // If the LHS has annotations, we perform an additional check for nullable value types
+                CheckDisallowedNullAssignment(rightState, leftAnnotations, right.Syntax.Location);
+
+                TrackNullableStateForAssignment(right, leftLValueType, MakeSlot(left), rightState, MakeSlot(right));
+                SetResult(node, TypeWithState.Create(leftLValueType.Type, rightState.State), leftLValueType);
             }
 
             return null;
         }
 
+        private FlowAnalysisAnnotations GetLValueAnnotations(BoundExpression expr)
+        {
+            var symbol = expr switch
+            {
+                BoundPropertyAccess property => property.PropertySymbol,
+                BoundIndexerAccess indexer => indexer.Indexer,
+                _ => null
+            };
+
+            var annotations = symbol switch
```

Comment:
>var annotations = symbol switch [](start = 12, length = 31)

`symbol` is already a `PropertySymbol` so this `switch` is just a `null` test. Consider simplifying.

For instance
```C#
var annotations = symbol?.SetMethod?.Parameters?.Last()?.FlowAnalysisAnnotations ??
    FlowAnalysisAnnotations.None;
```